// gerber-to-svg/lib/gerber-reader.js
(function() {
  var GerberReader;

  GerberReader = (function() {
    function GerberReader(gerberFile) {
      this.gerberFile = gerberFile != null ? gerberFile : '';
      this.line = 0;
      this.charIndex = 0;
      this.end = this.gerberFile.length;
    }

    GerberReader.prototype.nextBlock = function() {
      var char, current, parameter;
      if (this.index >= this.end) {
        return false;
      }
      current = '';
      parameter = false;
      if (this.line === 0) {
        this.line++;
      }
      while (!(this.charIndex >= this.end)) {
        char = this.gerberFile[this.charIndex++];
        if (char === '%') {
          if (!parameter) {
            parameter = [];
          } else {
            return {
              param: parameter
            };
          }
        } else if (char === '*') {
          if (parameter) {
            parameter.push(current);
            current = '';
          } else {
            return {
              block: current
            };
          }
        } else if (char === '\n') {
          this.line++;
        } else if ((' ' <= char && char <= '~')) {
          current += char;
        }
      }
      return false;
    };

    GerberReader.prototype.getLine = function() {
      return this.line;
    };

    return GerberReader;

  })();

  module.exports = GerberReader;

}).call(this);
